docs/proposal/20210823-gpu-support.md

```diff
@@ -0,0 +1,197 @@
+ // virtual machine is cloned.
+ // +optional
+ DeviceId int32 `json:"deviceId,omitempty"`
+
+ // VendorId is the vendor ID of a virtual machine's PCI, in integer.
+ // Defaults to the eponymous property value in the template from which the
+ // virtual machine is cloned.
+ // +optional
+ VendorId int32 `json:"vendorId,omitempty"`
+}
+
+```
+
+#### Controller Changes
+
+- If VGPU is present, VGPU's `ProfileName` would be fetched from `VirtualMachineCloneSpec` and device object (`createPCIPassThroughDevice` from govmomi library) is formed using this VGPU `ProfileName`. This device object is added to
+- the slice of devices and then used by `vspheremachine_controller` to schedule nodes.
+- If PCI device is present, PCI device's `DeviceId` and `VendorId` would be fetched from `VirtualMachineCloneSpec` and device object(`createPCIPassThroughDevice` from govmomi library) is formed using this information and then used by the `vspheremachine_controller` to schedule nodes.
+- The controller initiallizes variables respective to VGPU i.e `GPUProfileName`.
+- Similarly, controller also initializes variables related to Direct devices(for PCI passthrough and GPUDirect) ie. `GPUDeviceId` and `GPUVendorId`.
+- Using the above variables, respective objects of backing info of type `vimTypes.VirtualPCIPassthroughVmiopBackingInfo` with `{Vgpu: <GPUProfileName>}` for VGPU's and `vimTypes.VirtualPCIPassthroughAllowedDevice` for GPUDirect should be created by the controller.
+- Controller will call a function called as createPCI device which will use the information in above step to create objects of type `vimTypes.VirtualPCIPassthrough`
+- The create PCI device function will also be responsible for comparing the current devices with the expected devices.
+- The create method will return a slice `deviceChanges` of type `vimTypes.BaseVirtualDeviceConfigSpec`.
+- If there are no current devices and new devices are added. The slice `devicechanges` would reflect only add operations which would be equivalent to the number of devices added.
+- If the current devices are same as proposed devices. The length of the device changes remains 0.
+- Adding Vgpu devices and PCI devices with differnet backing info would return one remove and one add operation for each device.
+So the length of device changes would be 2*devices with different backing info.
+
+## Additional Details
+
+### Test Plan
+
+- Unit tests for cluster controller to test behavior when vGPU is requested.
+- Unit tests for cluster controller to test behavior if vGPU works as expected.
+- Unit tests for cluster controller to test if multiple vGPU's are requested, do all of them behave as expected.
+- Unit tests for cluster controller to test GPU's requested from GPU direct work as expected.
+- Unit tests for cluster controller to test PCI passthrough works as expected.
+- E2E test with a GPU Direct cluster running in different availability zones.
```

Comment:
Upstream CI does not have GPU support, so running E2E tests would not be possible.

Reply:
Need a broader discussion on this.

Reply:
Lets create an issue to track the E2E testing for GPU separately. Whatever mechanism we end up taking to run the tests on GPU specialized hardware hosted behind a VPN, we should use the issue to describe the solution and update the progress.

Reply:
Noted this one as a work item. Marking this resolved.